services/appveyor/appveyor.js

```diff
@@ -1,26 +1,23 @@
 'use strict';
 
 const BaseService = require('../base');
+const {
+  checkErrorResponse,
+  asJson,
+} = require('../../lib/error-helper');
 
-/**
- * AppVeyor CI integration.
- */
 module.exports = class AppVeyor extends BaseService {
   async handle({repo, branch}) {
     let apiUrl = 'https://ci.appveyor.com/api/projects/' + repo;
     if (branch != null) {
       apiUrl += '/branch/' + branch;
     }
-    const {buffer, res} = await this._sendAndCacheRequest(apiUrl, {
+    const json = await this._sendAndCacheRequest(apiUrl, {
       headers: { 'Accept': 'application/json' }
-    });
+    }).then(checkErrorResponse.asPromise({ notFoundMessage: 'project not found or access denied' }))
```

Comment:
I know this is already a good reduction in boilerplate, but I feel like we're very often going to chain `.then(checkErrorResponse.asPromise({}))` here. Do you think it would be worth maybe wrapping this in the base class so that the error handling function is passed as a param with a default value?

.. something like:

```js
  async _wrappedRequest(url, options, handler=checkErrorResponse.asPromise({})) {
    return this._sendAndCacheRequest(url, options).then(handler);
  }
```

in `base.js` and then we can call `const json = await this._wrappedRequest(url, options).then(asJson);` when we want to use the default handler (which is probably ~90% of cases) and `const json = await this._wrappedRequest(url, options, myCustomHander).then(asJson);` in the case where we want to customise the message or deal with a non-standard error condition or whatever.

If there is a situation where something _really_ non-standard happens, we could always override `_wrappedRequest` (or whatever we call it - better names welcome) or just call `_sendAndCacheRequest` directly. Does that seem sensible?

Reply:
Agreed, the vast majority of the time we're wanting to do the same thing here. Adding a method is a nice idea. Perhaps it could be `_performRequest` or `_makeRequest`.

For a custom not-found message, which is or should be relatively common, perhaps we could use a separate `static get notFoundMessage()` getter.

As you say, this could be overridden when something more custom is needed.

I'd even go so far as to say that, perhaps, the method should be `_requestJson`, which parses JSON by default, and the the handful of services which need something else could define or use a different method, or call external helpers (like the SVG badge fetcher). Then the 90% case looks like `const json = await this._requestJson(url, options);`, including if we're overriding the not-found message, and doesn't need to `require` any helpers.

I'd prefer to have concerns a little more separated – having JSON-defaulting stuff in BaseService begins to cross the line for me – but we can revisit that later. With a few more examples, probably separating concerns will be a bit easier.

Reply:
> I'd even go so far as to say that, perhaps, the method should be `_requestJson`
> ...
> I'd prefer to have concerns a little more separated – having JSON-defaulting stuff in BaseService begins to cross the line for me 

I had a very similar thought to this.

Maybe this is another good way to look at this:
Now `BaseService` is a class we can extend it, so you can declare

```js
class BaseJsonService extends BaseService
```

and chain `.then(asJson)` in the `BaseJsonService` version of whatever we call that method, but not in `BaseService`. If we spot other common patterns that we are using in lots of places, we could also define other base classes. We don't have to do it all in one object.

Reply:
Sure! I'm good with that as a starting point. Though, down the line, I'd also like to consider breaking the vendor-calling code out of the service class.